### ISSUE TITLE
Expose model discretization, particle operators, and local affine parameters

### ISSUE THREAD

**Issue by ma9o** [ma9o/dynestyx#4]
_Fork preview of the issue to file on BasisResearch/dynestyx._

(Cont'd from Discord)

I've been developing a framework, [nof1-causal-lab](https://github.com/ma9o/nof1-causal-lab), for LLM-driven SSM specification and visualisation, but haven't been able to bolt it directly onto Dynestyx and Cuthbert. The model side maps onto `DynamicalModel` one to one. The inference side does not: its estimation step is a joint parameter-and-trajectory particle MCMC that runs outside a NumPyro model, and what it needs from the model (a discrete transition it can both sample and score on the observation lattice, emission scores for rows with missing entries, and local affine parameters for a warm start) Dynestyx computes only inside `Discretizer` and the Cuthbert integrations. The conditional SMC kernels it needs are not in Cuthbert yet either. So the framework carries a private copy of all of it.

Specifically, I've implemented an LLM-facing specification layer, where the model proposes constructs, indicators, causal structure and priors under a state machine, and deterministic code compiles the result into a continuous-time nonlinear SDE with additive diffusion, heterogeneous emissions (Gaussian, Student-t, Poisson, negative binomial, Bernoulli, ordinal, categorical), controls, irregular observation times and partially missing rows. Estimation is a marginalized particle Gibbs after [Corenflos (2025)](https://arxiv.org/abs/2505.04611): a parameter-ensemble move (random walk or corrected pseudo-Langevin) alternating with a conditional parallel-in-time dSMC latent update ([Corenflos, Chopin and Särkkä 2022](https://arxiv.org/abs/2202.02264)) with exactly corrected leaf proposals, on Euler-Maruyama transitions. Gaussian approximations (IEKS, Pathfinder) only initialize the chain; every reported quantity goes through the exact engines. For this the framework re-derives from its spec the discrete transition (sample, score, and pairwise scores at dSMC block seams), masked emission scores and path scores, plus the kernels themselves, and several of the numerical mistakes I have caught in my own code were in that copy.

In [PR #3](https://github.com/ma9o/dynestyx/pull/3) I'm proposing to make public some of the internal seams of Dynestyx so that, for example, I could do:

```python
import dynestyx as dsx
from dynestyx.inference.configs.discretizer import EulerMaruyamaConfig, LocalLinearizationConfig

# the transition the sampler both samples and scores, on the observation lattice
ops = dsx.build_particle_operators(dsx.discretize_dynamics(model, EulerMaruyamaConfig()))
interval = dict(previous_control=u, previous_time=t0, time=t1)

x = ops.transition_sample(key, x_prev, **interval)
ops.transition_log_prob(x_prev, x, **interval)
ops.pairwise_transition_log_prob(left_block, right_block, **interval)  # dSMC block seams
ops.observation_log_prob(x, observation=y, control=None, time=t1, observation_mask=~jnp.isnan(y))

# local affine parameters for the Gaussian warm start
params = dsx.linearized_transition_parameters(
    model, LocalLinearizationConfig(), linearization_state=x_ref, **interval
)
```

Everything there already exists inside Dynestyx; the PR moves it behind public names and makes `Discretizer`, the Cuthbert PF and the backward smoother its first consumers.

In the future it would also be useful to:

1. have conditional trajectory kernels in Cuthbert (CSMC and conditional dSMC) behind one `step(key, reference_path)` interface, continuing state-space-models/cuthbert#181. They would take the sample and score operators above, and the pairwise transition score is what the dSMC seams need. I'd raise that there.
2. build a joint parameter/trajectory kernel on top of them in Cuthbert, following Corenflos (2025).
3. add a small runtime to Dynestyx that composes a NumPyro prior program (transforms, prior replay) with these operators and drives that kernel, with control indexing and missingness explicit.

With those in place nof1 keeps its model adapter and reporting policy and deletes its runtime bundle and kernel code. I have prototyped all of it end to end on my forks ([cuthbert](https://github.com/ma9o/cuthbert/tree/simulation/joint-particle-kernel), [dynestyx](https://github.com/ma9o/dynestyx/tree/simulation/particle-runtime)) and would rather land it one piece at a time, starting with the PR above. Further out, the generic pieces nof1 still owns, heterogeneous emission composition and the IEKS/Pathfinder initialization helpers, could follow the same route.
